torch/csrc/inductor/math.h

```diff
@@ -0,0 +1,57 @@
+#pragma once
+
+#include <cstdint>
+#include <cmath>
+#include "c10/macros/Macros.h"
+
+// Functions in this file should be header-only as it is used in ABI-compatible mode.
```

Comment:
I don't understand. Is this OK to run from the DSO? Why does it include Macros.h then?

Reply:
Macros.h doesn't include anything that links to external C++ code, so it's safe

Reply:
Your statement is true, but it also felt like a slippery slope to me. People who touch those recursively included heads won't necessarily know the ABI constraint of AOTI. Say someone decides to split a header into .h and .cpp, and then it will break our assumption here.

With that being said, I am ok with the current approach if you can also update the linker flags for the abi_compatible mode, e.g.  https://github.com/pytorch/pytorch/blob/cb233dada43f016c8fa58979838939c1adda416b/torch/_inductor/codecache.py#L1301C24-L1301C24. Then if the aforementioned change happens, at least we should be able to catch it with our abi_compatible test cases.

Reply:
Hmm, on a second thought, my suggestion may not be trivial to implement, because the C shim layout does live in libtorch so we need to link against that.

Reply:
> Your statement is true, but it also felt like a slippery slope to me.

Yes, I agree that the situation is not ideal. That said I don't think I am actually making the situation that much worse, because cpp_prefix.h already includes a bunch of c10 and aten headers directly (see updated message at the top of the PR.)

> With that being said, I am ok with the current approach if you can also update the linker flags for the abi_compatible mode

It looks like we already avoid linking against c10 when building in fbcode, so I think we're good? I guess it would be better if we checked for `abi_compatible` instead of `is_fbcode`, but it seems they are being used interchangeably throughout the codebase: https://github.com/pytorch/pytorch/blob/cb233dada43f016c8fa58979838939c1adda416b/torch/_inductor/codecache.py#L1398-L1399

(Side note: we are commenting on the "make an alternative implementation of `div_floor_*` functions" approach, which is not the latest version of this PR; the latest version just relocates the existing `div_floor_*` functions from ATen to c10. Either way, the non-linking of c10 makes this safe.)

Reply:
> pp_prefix.h already includes a bunch of c10 and aten headers directly

This should be fixed.

> I guess it would be better if we checked for abi_compatible instead of is_fbcode, but it seems they are being used interchangeably throughout the codebase...

The long-term goal is to only keep the abi-compatible mode for OSS as well. In this case, please change it to check `abi_compatible